### PR TITLE
[node] improve typing of query.parse()

### PR DESCRIPTION
**Pull request by Flarna**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/dist/latest-v9.x/docs/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Followup on #20720. Try to further improve typing of `url.parse()` and to avoid reported problems.
@connor4312 @bowdenk7 Please take a look, try in your code and let us know the results. Thanks!

### CODE REVIEW

**Review by connor4312 — APPROVED** [review]

File: types/node/index.d.ts [hunk 1]
```diff
@@ -2342,7 +2342,19 @@ declare module "url" {
         query?: string | null | ParsedUrlQuery;
     }
 
-    export function parse(urlStr: string, parseQueryString?: boolean, slashesDenoteHost?: boolean): Url;
+    export interface UrlParsedQuery extends Url {
+        query: ParsedUrlQuery;
+    }
+
+    export interface UrlStringQuery extends Url {
```
Comment (OliverJAsh) [inline types/node/index.d.ts]: Perhaps rename these types to `UrlWith{StringQuery,ParsedQuery}` for readability?

File: types/node/index.d.ts [hunk 2]
```diff
@@ -2342,7 +2342,19 @@ declare module "url" {
         query?: string | null | ParsedUrlQuery;
```
Comment (OliverJAsh) [inline types/node/index.d.ts]: Maybe remove this, and make the `boolean` `parse` overload return `UrlParsedQuery | UrlStringQuery`?

**Review by OliverJAsh — APPROVED** [review]

**Comment by Flarna** [comment]
Thanks for the review. Pushed an update.

File: types/node/index.d.ts (hunk 2 above)
Comment (Flarna) [reply types/node/index.d.ts]: seems `Url.url` is used at several other modules so I will keep it.

**Review by OliverJAsh — APPROVED** [review]

**Comment by Flarna** [comment]
CI fails unrelated, will try to fix via another PR but it seems fails are introduced faster then I can fix them...

**Comment by Flarna** [comment]
@mhegazy, @andy-ms restify fail is unrelated, created #22653 to correct it.
Any chance to get this merged before again some unrelated change breaks build?

**Pull request merged**